File: clients/javascript/open-feature-provider/examples/local-with-fallback-example.ts
/**
 * Local with Fallback Example
 *
 * Demonstrates using LocalResolutionProvider with HTTP primary source
 * and File fallback source. If the HTTP server is unavailable, the
 * provider falls back to reading from a local config file.
 *
 * Usage:
 *   npm run build
 *   node dist/examples/local-with-fallback-example.js
 */

import { OpenFeature } from "@openfeature/server-sdk";
import { HttpDataSource, FileDataSource } from "../data-sources";
import { LocalResolutionProvider } from "../local-resolution-provider";
import * as path from "path";

async function sleep(ms: number): Promise<void> {
    return new Promise((resolve) => setTimeout(resolve, ms));
}

async function main() {
    console.log("=== Superposition Fallback + Polling Example ===\n");

    // Create HTTP data source (primary)
    const httpSource = new HttpDataSource({
        endpoint: process.env.SUPERPOSITION_ENDPOINT || "http://localhost:8080",
        token: process.env.SUPERPOSITION_TOKEN || "token",
        orgId: process.env.SUPERPOSITION_ORG_ID || "localorg",
        workspaceId: process.env.SUPERPOSITION_WORKSPACE_ID || "dev",
    });

    // Create File data source (fallback)
    const configPath = path.join(__dirname, "config.toml");
    console.log(`Primary: HTTP (${httpSource.endpoint})`);
    console.log(`Fallback: ${configPath}`);
    console.log(`Polling every 10s. Printing currency and timeout values (Ctrl-C to stop).\n`);

    const fileSource = new FileDataSource(configPath);

    // Create provider with both sources and polling strategy
    const provider = new LocalResolutionProvider(
        httpSource,
        fileSource, // fallback
        { type: "polling", interval: 10 }
    );

    try {
        // Register with OpenFeature and create a client
        await OpenFeature.setProviderAndWait(provider);
        const client = OpenFeature.getClient();

        // Allow time for the provider to initialize via OpenFeature
        await sleep(2000);

        const context = {
            targetingKey: "user-456",
            os: "linux",
            city: "Berlin",
        };

        console.log("Starting polling loop...\n");

        // Poll for config changes and print values in a loop
        let iteration = 0;
        while (iteration < 12) {
            // Run for 1 minute (12 * 5s)
            const timestamp = new Date().toISOString().split("T")[1].split(".")[0];

            try {
                const currency = await client.getStringValue(
                    "currency",
                    "USD",
                    context
                );
                process.stdout.write(`[${timestamp}] currency = ${currency}`);
            } catch (e) {
                process.stdout.write(`[${timestamp}] currency error: ${e}`);
            }

            try {
                const timeout = await client.getNumberValue("timeout", 30, context);
                console.log(`  |  timeout = ${timeout}`);
            } catch (e) {
                console.log(`  |  timeout error: ${e}`);
            }

            await sleep(5000);
            iteration++;
        }

        console.log("\nExample completed!");
    } catch (error) {
        console.error("Error:", error);
    } finally {
        // Cleanup
        await provider.onClose();
    }
}

main();

File: clients/java/openfeature-provider/src/main/java/io/juspay/superposition/examples/LocalWithFallbackExample.java
package io.juspay.superposition.examples;

import dev.openfeature.sdk.*;
import io.juspay.superposition.client.SuperpositionAsyncClient;
import io.juspay.superposition.client.auth.BearerTokenIdentityResolver;
import io.juspay.superposition.provider.*;
import software.amazon.smithy.java.client.core.endpoint.EndpointResolver;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.TimeUnit;

/**
 * Local with Fallback Example
 *
 * Demonstrates using LocalResolutionProvider with HTTP primary source
 * and File fallback source. If the HTTP server is unavailable, the
 * provider falls back to reading from a local config file.
 *
 * Usage:
 *   ./gradlew :openfeature-provider:run -PmainClass=io.juspay.superposition.examples.LocalWithFallbackExample
 */
public class LocalWithFallbackExample {

    public static void main(String[] args) throws Exception {
        System.out.println("=== Superposition Fallback + Polling Example ===\n");

        // Get configuration from environment variables
        String endpoint = System.getenv().getOrDefault("SUPERPOSITION_ENDPOINT", "http://localhost:8080");
        String token = System.getenv().getOrDefault("SUPERPOSITION_TOKEN", "token");
        String orgId = System.getenv().getOrDefault("SUPERPOSITION_ORG_ID", "localorg");
        String workspaceId = System.getenv().getOrDefault("SUPERPOSITION_WORKSPACE", "dev");

        // Create HTTP data source (primary)
        var sdkClient = SuperpositionAsyncClient.builder()
            .endpointResolver(EndpointResolver.staticEndpoint(endpoint))
            .addIdentityResolver(new BearerTokenIdentityResolver(token))
            .build();

        HttpDataSource httpSource = new HttpDataSource(sdkClient, orgId, workspaceId);

        // Create File data source (fallback)
        Path configPath = Paths.get("src/main/resources/config.toml");
        System.out.println("Primary: HTTP (" + endpoint + ")");
        System.out.println("Fallback: " + configPath);
        System.out.println("Polling every 10s. Printing config every 5s (Ctrl-C to stop).\n");

        // For this example, we'll use HTTP as both primary and fallback
        // In a real scenario, you would use FileDataSource as fallback
        HttpDataSource fallbackSource = new HttpDataSource(sdkClient, orgId, workspaceId);

        // Create provider with both sources and polling strategy
        LocalResolutionProvider provider = new LocalResolutionProvider(
            httpSource,
            Optional.of(fallbackSource),
            new RefreshStrategy.Polling(10),
            Optional.empty()
        );

        try {
            // Register with OpenFeature and create a client
            OpenFeatureAPI api = OpenFeatureAPI.getInstance();
            EvaluationContext initContext = new ImmutableContext(Map.of());
            provider.initialize(initContext);
            api.setProvider(provider);
            Client client = api.getClient();

            // Allow time for the provider to initialize
            TimeUnit.SECONDS.sleep(2);

            EvaluationContext context = new ImmutableContext(Map.of(
                "os", new Value("linux"),
                "city", new Value("Berlin")
            ));

            System.out.println("Starting polling loop...\n");

            // Poll for config changes and print values in a loop
            for (int i = 0; i < 12; i++) {
                String timestamp = java.time.LocalTime.now().toString().substring(0, 8);

                try {
                    String currency = client.getStringValue("currency", "USD", context);
                    System.out.print("[" + timestamp + "] currency = " + currency);
                } catch (Exception e) {
                    System.out.print("[" + timestamp + "] currency error: " + e.getMessage());
                }

                try {
                    int timeout = client.getIntegerValue("timeout", 30, context);
                    System.out.println("  |  timeout = " + timeout);
                } catch (Exception e) {
                    System.out.println("  |  timeout error: " + e.getMessage());
                }

                TimeUnit.SECONDS.sleep(5);
            }

            System.out.println("\nExample completed!");

        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        } finally {
            // Cleanup
            provider.shutdown();
        }
    }
}
